lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -173,8 +192,55 @@ impl Source for ChecksSource {
             }
         }
 
+        listener_shutdown_coordinator.shutdown().await;
+
         info!("Checks source stopped.");
 
         Ok(())
     }
 }
+
+async fn drain_and_dispatch_check_metrics(
+    shutdown_handle: DynamicShutdownHandle, context: SourceContext, mut check_metrics_rx: mpsc::Receiver<CheckMetric>,
+) {
+    tokio::pin!(shutdown_handle);
+
+    loop {
+        select! {
+            _ = &mut shutdown_handle => {
+                debug!("Checks events listerners received shutdown signal.");
+                break;
+            }
+            result = check_metrics_rx.recv() => match result {
+                None => break,
+                Some(check_metric) => {
+                    let mut buffered_dispatcher = context
+                        .dispatcher()
+                        .buffered()
+                        .expect("checks source should always have default output");
+
+                    trace!("Received check metric: {:?}", check_metric);
+                    let event: Event = check_metric.into();
+
+                    if let Err(e) = buffered_dispatcher.push(event).await {
+                        error!(error = %e, "Failed to forward check metrics.");
+                    }
+
+                    if let Err(e) = buffered_dispatcher.flush().await {
+                        error!(error = %e, "Failed to flush check metrics.");
+                    }
+                }
+            }
+        }
+    }
+
+    let buffered_dispatcher = context
+        .dispatcher()
+        .buffered()
+        .expect("checks source should always have default output");
+
+    // Final dispatcher flush after all checks have closed.
+    if let Err(e) = buffered_dispatcher.flush().await {
+        error!(error = %e, "Failed to flush check metrics.");
+    }
```

Comment:
No need for this: you're constructing it fresh so it's only ever going to be empty.